Add support for `showPicker()` to `HTMLSelectElement`

lukewarlow opened the pull request:
#### 5df7cdc4a1ec1cfef6b892665536af6eefe925e2
<pre>
Add support for `showPicker()` to `HTMLSelectElement`
<a href="https://bugs.webkit.org/show_bug.cgi?id=261814">https://bugs.webkit.org/show_bug.cgi?id=261814</a>

Reviewed by Aditya Keerthi.

At whatwg/html#9754, it&apos;s proposed to add a
new HTMLSelectElement::showPicker() method to show the popup for
select elements.

This patch introduces this functionality, for non-iOS platforms in WK2.

* LayoutTests/fast/forms/select/select-show-picker-expected.txt: Added.
* LayoutTests/fast/forms/select/select-show-picker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-cross-origin-iframe.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-disabled.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-multiple.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-size.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-user-gesture.tentative-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::showPicker):
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLSelectElement.idl:
* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:

Canonical link: <a href="https://commits.webkit.org/269420@main">https://commits.webkit.org/269420@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/803aebbd3f94149a192e2a15bcc52028c65ba19e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/22492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/143 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/23571 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/24397 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20821 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22750 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/1220 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/23024 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/21821 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22732 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/148 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19533 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/25249 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/114 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/20380 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26637 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/19596 "Built successfully and passed tests") | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20633 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/24478 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/21861 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/80 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17945 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/25910 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/59 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/20188 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6083 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5360 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/98 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/27189 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/95 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/5910 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

lukewarlow commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1658,4 +1659,33 @@ unsigned HTMLSelectElement::length() const
+        return Exception { NotAllowedError, "Select showPicker() requires a user gesture."_s };
+
+#if PLATFORM(IOS_FAMILY)
+    focus();
```
The showPicker spec isn't clear on if this is allowed or not (see https://github.com/whatwg/html/issues/9757) but given the spirit of the feature it feels like working rather than not working is the best solution.

lukewarlow replied:
Moving this into a follow up that addresses showPicker on iOS in general

pxlcoder commented on LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-cross-origin-iframe.tentative-expected.txt:
```diff
@@ -0,0 +1,10 @@
+PASS Test showPicker() called from cross-origin iframe
+PASS Test showPicker() called from cross-origin iframe 1
+PASS Test showPicker() called from cross-origin iframe 2
+TIMEOUT Test showPicker() called from cross-origin iframe 3 Test timed out
```
Curious what's up with this. Not that it needs to be resolved prior to merging this change.

lukewarlow replied:
Yeah I'm not really sure the same happens for input showPicker too

pxlcoder commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -5838,6 +5838,20 @@ SecureContextChecksEnabled:
+  status: testable
+  category: dom
+  humanReadableName: "Select showPicker() method"
+  humanReadableDescription: "Enable showPicker() method on select"
```
Nit: "Select" -> "<select>".

lukewarlow replied:
Sorry can you clarify what you mean by this?

lukewarlow replied:
Do you mean "humanReadableName: "showPicker() method""? That would be confusing because it doesn't also control the input.showPicker() method?

pxlcoder replied:
I mistyped sorry, I meant to say "Select" -> `<select>`.

(edit: I had typed `"<select>"`, but that just became in `""` in Markdown.)

lukewarlow replied:
Ah that makes sense, GitHub converting that always catches me out too.

pxlcoder commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1667,4 +1668,31 @@ unsigned HTMLSelectElement::length() const
+#if !PLATFORM(IOS_FAMILY)
+    auto* renderer = this->renderer();
+    if (is<RenderMenuList>(renderer))
+        downcast<RenderMenuList>(*renderer).showPopup();
```
```suggestion
    if (auto* renderMenuList = dynamicDowncast<RenderMenuList>(renderer))
        renderMenuList->showPopup();
```

pxlcoder commented on Source/WebCore/html/HTMLSelectElement.cpp:
```diff
@@ -1667,4 +1668,31 @@ unsigned HTMLSelectElement::length() const
 
+ExceptionOr<void> HTMLSelectElement::showPicker()
+{
+    auto* frame = document().frame();
```
`RefPtr frame = ` as we are generally trying to adopt smart pointers everywhere.

pxlcoder reviewed (commented):
For this one, I think "Add support for `showPicker()` to `<select>``" would be a better title.

pxlcoder reviewed (approved):
Looks good, following the update to the setting name/description.

lukewarlow committed:
Add support for `showPicker()` to `HTMLSelectElement`
https://bugs.webkit.org/show_bug.cgi?id=261814

Reviewed by Aditya Keerthi.

At whatwg/html#9754, it's proposed to add a
new HTMLSelectElement::showPicker() method to show the popup for
select elements.

This patch introduces this functionality, for non-iOS platforms in WK2.

* LayoutTests/fast/forms/select/select-show-picker-expected.txt: Added.
* LayoutTests/fast/forms/select/select-show-picker.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-cross-origin-iframe.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-disabled.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-multiple.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-size.tentative-expected.txt: Added.
* LayoutTests/imported/w3c/web-platform-tests/html/semantics/forms/the-select-element/show-picker-user-gesture.tentative-expected.txt: Added.
* LayoutTests/platform/ios/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/html/HTMLSelectElement.cpp:
(WebCore::HTMLSelectElement::showPicker):
* Source/WebCore/html/HTMLSelectElement.h:
* Source/WebCore/html/HTMLSelectElement.idl:
* LayoutTests/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:

Canonical link: https://commits.webkit.org/269420@main
